lib/handlerTemplate.js

```diff
@@ -164,151 +128,337 @@ function validateSpec(spec) {
+    var lastStep = spec[spec.length - 1];
+    if (countReturnsInStep(lastStep) === 0) {
+        if (Object.keys(lastStep).length > 1) {
+            throw new Error('Invalid spec. Need a return if the last step is parallel.');
+        } else {
+            // Make the last step a return step.
+            var lastRequestName = Object.keys(lastStep)[0];
+            lastStep[lastRequestName].return = true;
+        }
+    }
+}
+
+function courteousExpand(template, ctx, info) {
+    try {
+        return template.expand(ctx.model);
+    } catch (e) {
+        e.reqName = info.reqName;
+        e.reqSpec = info.spec;
+        throw e;
     }
 }
 
 /**
- * Prepares a request chain on startup: compiles request/response templates,
- * creates catch conditions verifiers.
- * @param {object} spec request chain configuration
- * @param {object} conf, an optional config object to be merged into the
- *                  globals available in template expressions.
- *
- * @returns {Array} an array of prepared steps, each containing a
- *                  request function generator
+ * Creates a request handler.
+ * @param info, info about a request.
+ *          - {string name: request name
+ *          - {object} spec: object containing a request template
+ * @return
  */
-function prepareRequestChain(spec, conf) {
-
-    /**
-     * Creates a function generator that returns a closure to execute in order to
-     * make a request
-     * @param requestSpec request spec object, the result would be added there under
-     *        a request property
-     * @param requestName a name of the request, the result would be added to the
-     *        context under this name
-     * @param requestConf request configuration object containing a request template
-     */
-    function prepareRequest(requestSpec, requestName, requestConf) {
+function makeRequestHandler(info, options) {
+    if (info.spec.request) {
         var template;
-        if (requestConf.request) {
-            template = new Template(requestConf.request, conf);
-            requestSpec.request = function(restbase, context) {
-                var req = template.expand(context);
-                if (!requestConf.request.method) {
-                    // Do we really want this?
-                    req.method = context.request.method || 'get';
+        try {
+            template = new Template(info.spec.request, options.globals);
+        } catch (e) {
+            e.requestName = info.name;
+            e.requestSpec = info.spec.spec;
+            e.message = 'Template compilation failed. See .spec for details. ' + e.message;
+            throw e;
+        }
+
+        var catchPred;
+        if (info.spec.catch) {
+            catchPred = compileCatchFunction(info.spec.catch);
+        }
+
+        var shouldReturn;
+        // Important: `return_if` takes precedence over return, so that a
+        // `return` with `return_if` present behaves like `response`.
```

Comment:
The template is actually not applied either if the condition doesn't evaluate to true. Doing so would cause failures for typical try-storage use cases that catch 404s, which don't provide the information expected in the return template. The lack of clear semantics around this is one of the less-than-ideal consequences of separating return_if from return.

An idea we came up with during our conversation this morning is to leverage our new templating powers to support something like

```yaml
return:
  if: '{{status_match(this_request.status, '2xx')}}'
  status: 200
  body: 
   ...
```
or, without returning,
```yaml
response:
  if: '{{status_match(this_request.status, '2xx')}}'
  status: 200
  body: 
   ...
```